src/analyze.rs

```diff
@@ -0,0 +1,80 @@
+use std::{
+    fs::File,
+    io::{Read, Seek},
+    path::Path,
+    thread::sleep,
+    time::Duration,
+};
+
+use rodio::{buffer::SamplesBuffer, Decoder, OutputStream, Source};
+
+use crate::{base::Res, listen::get_notes_from_audio_data, note::Note};
+
+/// Retrieve a list of notes which are guessed from the given audio clip.
+pub fn get_notes_from_audio_file(file: impl AsRef<Path>, start: Option<Duration>, end: Option<Duration>) -> Res<Vec<Note>> {
+    let path = file.as_ref();
+    let file = File::open(path)?;
+    let start = start.unwrap_or_default();
+    let decoder = Decoder::new(file)?;
+    let decoder = decoder.skip_duration(start).convert_samples();
+    let sample_rate = decoder.sample_rate() as usize;
+    let samples: Vec<_> = if let Some(end) = end { decoder.take_duration(end - start).collect() } else { decoder.collect() };
```

Comment:
Go ahead and spell this out with carriage returns?

Reply:
I'm not sure what you mean. This is just following the formatting rules?

Reply:
Ahhh, damn.  Didn't know the formatting would change that.

Ok, this is fine.  Add some vertical white space to group operations?